src/ray/common/file_system_monitor.h

```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <gtest/gtest_prod.h>
+
+#include <filesystem>
+#include <optional>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/asio/periodical_runner.h"
+
+namespace ray {
+/// Monitor the filesystem capacity ray is using.
+/// This class is thread safe.
+class FileSystemMonitor {
+ public:
+  /// Constructor.
+  ///
+  /// \param paths paths of the file system to monitor the usage.
+  /// \param capacity_threshold a value between 0-1 indicates the capacity limit.
+  /// \param monitor_interval_ms control the frequency to check the disk usage.
+  FileSystemMonitor(std::vector<std::string> paths,
+                    double capacity_threshold,
+                    uint64_t monitor_interval_ms = 1000);
+
+  ~FileSystemMonitor();
+
+  /// return the disk usage.
+  std::optional<std::filesystem::space_info> Space(const std::string &path) const;
+
+  /// returns true if ANY path's disk usage is over the capacity threshold.
```

Comment:
Does this mean we will raise an exception although there are disks that are not over capacity?

Reply:
that's right. I think in practice it might be fine, since most of the time the spill path and fallback path is the same device. 
we can make this logic more complicated, if the current spill logic could avoid disk full path if multiple spill folders are provided.